File: library.py
class Library:
    
    def __init__(self):
        self.i = 0 
        self.books = dict()

    def find_book(self, title = None, writer = None):
        if title:
            for book,obj in self.books.items():
                if obj["title"] == title:
                    return book
            return 0
        elif writer:
            for book,obj in self.books.items():
                if obj["writer"] == writer:
                    return book
            return 0
        else:
            return 0

    def read_book(self, title = None, writer = None):
        if self.find_book(self, title = None, writer = None) == 0:
            print("We could not find the book. Please try again...")
        else:
            id = self.find_book(self, title, writer)
            publish_date = self.books[id].get("publish_date")
            summary = self.books[id].get("summary")
            if title:
                writer = self.books[id].get("writer")
            elif writer:
                self.books[id].get("title")
            print("Here are the information about the book")
            print("Title: " + title)
            print("Writer: " + writer)
            print("Date published: " + publish_date)
            print("Summary: " + summary)
            
    def delete_book(self, title = None, writer = None):
        if self.find_book(self, title = None, writer = None) == 0:
            return 0
        else:
            self.books.pop(id)
            return 1
    
    def update_book(self,id):
        print("Do you want to change the title? yes/no")
        answer = input()
        if answer == "yes":
            print("Enter title!")
            title= input()
            self.books[id].update({"title": title})
        print("Do you want to change the writer? yes/no")
        if answer == "yes":
            print("Enter writer!")
            writer = input()
            self.books[id].update({"writer": writer})
        print("Do you want to change the date of publishment? yes/no")
        if answer == "yes":
            print("Enter date!")
            publish_date = input()
            self.books[id].update({"publish_date": publish_date})
        print("Do you want to change the summary? yes/no")
        if answer == "yes":
            print("Enter summary!")
            summary = input()
            self.books[id].update({"summary": summary})


    def add_book(self):
        self.i = self.i+1
        id = self.i
        print("Add the title of the book:")
        title = input()
        print("Add the writer of the book:")
        writer = input()        
        print("Add the date that the book was published of the book:")
        publish_date = input()
        print("Add the summary of the book:")
        summary = input()
        self.books[id]= {"title": title, 
                         "writer" : writer, 
                         "publish_date" : publish_date, 
                         "summary" : summary
                        }

    def menu(self):
        flag = True
        while flag:
            print("The menu is:")
            print("1. Read book")
            print("2. Add a new book")
            print("3. Update a book")
            print("4. Delete a book")
            print("5. exit")
            print("Please press a number!!")
            number = input()
            if isinstance(number,int) and number == 1:
                print("Please enter the 1 to enter the title or 2 to enter the writer of the book")
                nu = int(input())
                if isinstance(number,int) and number == 1:
                    print("Please enter the title of the book:")
                    title = input()
                    self.read_book(self, title, None)
                elif isinstance(number,int) and number == 2:
                    print("Please enter the writer of the book:")
                    writer = input()
                    self.read_book(self, None, writer)
                else:
                    print("There has been an error. Please try again!")
                    self.menu(self)
            elif isinstance(number,int) and number == 2:
                self.add_book(self)
            elif isinstance(number,int) and number == 3:
                print("Please enter the 1 to enter the title or 2 to enter the writer of the book you want to update")
                nu = int(input())
                if isinstance(number,int): 
                    if number == 1:
                        print("Please enter the title of the book:")
                        title = input()
                        id = self.find_book(self, title, None)
                    elif number == 2:
                        print("Please enter the writer of the book:")
                        writer = input()
                        id = self.find_book(self, None, writer)
                    if id == 0:
                        print("There has been an error. Please try again!")
                        self.menu(self)
                    else:
                        self.update_book(self,id)
                else:
                    print("There has been an error. Please try again!")
                    self.menu(self)
            elif isinstance(number,int) and number == 4:
                print("Please enter the 1 to enter the title or 2 to enter the writer of the book")
                nu = int(input())
                if isinstance(number,int) and number == 1:
                    print("Please enter the title of the book:")
                    title = input()
                    if ( self.delete_book(self, title, None) == 0) :
                        print("Could not delete the book!")
                    else:
                        print("Book deleted succesfully!")
                elif isinstance(number,int) and number == 2:
                    print("Please enter the writer of the book:")
                    writer = input()
                    self.delete_book(self, None, writer)
                else:
                    print("There has been an error. Please try again!")
                    self.menu(self)
            elif isinstance(number,int) and number == 5:
                print("Exiting...")
                flag = False
            else:
                print("There has been an error. Please try again!")
                self.menu(self) 


library = Library()
library.menu()

File: Library.java
import java.util.Scanner;
import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;

class Library {

	private HashMap<Integer,Book> books;
	private int counter;

	public Library() {
        // Initialize the list of books
        this.books = new HashMap<>();
        this.counter = 0 ;
    }

	//returns 1 if succesfull, otherwise returns 0
	public int addBook() {
		Scanner input = new Scanner(System.in);
		try {
			System.out.println("Please enter the title of the book:");
			String title = input.nextLine();
			System.out.println("Add the writer of the book:");
			String writer = input.nextLine();
			System.out.println("Add the year that the book was published of the book:");
			int publishYear = input.nextInt();
			input.nextLine();
			System.out.println("Add the summary of the book:");
			String summary = input.nextLine();
			counter++;
			int id = counter;
			Book book = new Book(title, writer, publishYear,summary);
			books.put(id,book);
			return 1;
		} catch (Exception e) {
            return 0;
        }
	}

	public void readBook(String title, String writer) {
		int bookId = findBook(title, writer);
		if (bookId == 0) {
			System.out.println("We could not find the book. Please try again!");
		} else {
			Book book = books.get(bookId);
			System.out.println("Title: " + book.getTitle());
			System.out.println("Author: " + book.getWriter());
			System.out.println("Year o the publishment: " + book.getYearPublished());
			System.out.println("Summary: " + book.getSummary());
		}
	}

	//returns 1 if succesfull, otherwise returns 0
	public int deleteBook(String title, String writer) {
		int bookId = findBook(title, writer);
		if (bookId == 0) {
			return 0;
		} else {
			books.remove(bookId);
			return 1;
		}
	}

	//returns 1 if succesfull, otherwise returns 0
	public int updateBook(String title, String writer) {
		Scanner input = new Scanner(System.in);
		int bookId = findBook(title, writer);
		if (bookId == 0) {
			return 0;
		} else {
			try {
				Book info = books.get(bookId);
				System.out.println("Do you want to change the title? yes/no");
				String answer = input.nextLine();
				if (answer == "yes") {
					System.out.println("Enter title!");
					title = input.nextLine();
					info.setTitle(title);
					books.put(bookId, info);
				}
				System.out.println("Do you want to change the author? yes/no");
				answer = input.nextLine();
				if (answer == "yes") {
					System.out.println("Enter author!");
					writer = input.nextLine();
					info.setWriter(writer);
					books.put(bookId, info);
				}
				System.out.println("Do you want to change the year of publishment? yes/no");
				answer = input.nextLine();
				if (answer == "yes") {
					System.out.println("Enter year!");
					int yearPublished = input.nextInt();
					info.setYearPublished(yearPublished);
					books.put(bookId, info);
				}
				System.out.println("Do you want to change the summary? yes/no");
				answer = input.nextLine();
				if (answer == "yes") {
					System.out.println("Enter summary!");
					String summary = input.nextLine();
					info.setSummary(summary);
					books.put(bookId, info);
				}
			} catch (Exception e) {
				return 0;
			}
			return 1;
		}

	}

	// returns bookId
	public int findBook(String title, String writer) {
		for (Map.Entry<Integer, Book> set :
		             books.entrySet()) {
			Book info = set.getValue();
			if (title != null) {
				if (info.getTitle().equals(title)) {
					return set.getKey();
				}
			} else if (writer != null) {
				if (info.getWriter().equals(writer)) {
					return set.getKey();
				}
			} else {
				return 0;
			}
		}
		return 0;
	}

}
